Dockerfile

```diff
@@ -1,31 +1,40 @@
-FROM quay.io/opendatahub/text-generation-inference:stable-bafd218
-
-USER root
+ARG BASE_UBI_IMAGE_TAG=8.8-1067
+FROM registry.access.redhat.com/ubi8/ubi:${BASE_UBI_IMAGE_TAG} as base
 
 # Add grpc-ecosystem health probe
 ARG GRPC_HEALTH_PROBE_VERSION=v0.4.20
 
-WORKDIR /caikit
-COPY caikit /caikit
-
-RUN yum -y update && yum -y install git git-lfs && yum clean all && \
-    git lfs install && \
-    pip install 'micropipenv[toml]' && \
-    micropipenv install && \
-    rm -rf ~/.cache && \
-    mkdir -p /opt/models && \
-    adduser -g 0 -u 1001 caikit --home-dir /caikit && \
-    chown -R 1001:0 /caikit /opt/models && \
-    chmod -R g=u /caikit /opt/models
-
 # This is for the use-cases without kserve
 RUN curl -Lo /usr/local/bin/grpc_health_probe https://github.com/grpc-ecosystem/grpc-health-probe/releases/download/${GRPC_HEALTH_PROBE_VERSION}/grpc_health_probe-linux-amd64 && \
     chmod +x /usr/local/bin/grpc_health_probe
 
-USER 1001
+RUN dnf -y update && \
+    dnf -y install \
+        git git-lfs python39-pip python39-wheel && \
+    git lfs install && \
+    pip3 install --no-cache-dir --upgrade pip wheel && \
+    dnf clean all
+
+
+FROM base as poetry-builder
+ENV POETRY_VIRTUALENVS_IN_PROJECT=1
+
+WORKDIR /tmp/poetry
+COPY pyproject.toml .
+COPY poetry.lock .
+RUN pip3 install poetry && poetry install
+
+
+FROM base as deploy
+WORKDIR /caikit
+
+COPY --from=poetry-builder /tmp/poetry/.venv /opt/caikit/
+
+ENV VIRTUAL_ENV=/opt/caikit
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
+
+RUN adduser -g 0 -u 1001 caikit --home-dir /caikit && chown -R caikit /caikit
```

Comment:
2 cents for improve the user addition:

```
groupadd -r caikit -g 1001 && useradd -u 1001 -r -g root -G "caikit" -m -d "/caikit" -s /sbin/nologin -c "Caikit User  user" caikit
```

ps. this runs well with ocp random uid as well.

Reply:
Any particular reason for having `-g root`?

Reply:
better readability, it is the same than `-g 0`

Reply:
There's some openshift-specific semantics around this, I think the symbolic name works, but the group 0 assigned I don't think is quite the same as the group that root is a part of, there's a doc about this somewhere with the exact behavior on openshift

Reply:
> better readability, it is the same than -g 0

Sorry, I meant if there's any specific reason for adding the `caikit` user to the root group, maybe it's just better to get rid of it now that I see it.

Reply:
@dtrifiro, it's the solution to support OpenShift arbitrary user ID, see there:
> https://docs.openshift.com/container-platform/4.11/openshift_images/create-images.html#use-uid_create-images
```
RUN chgrp -R 0 /some/directory && \
    chmod -R g=u /some/directory
```
that's the way to give read/write access to `anyuid` user, contrary to the more `chown 1001`  which implies that the container runs with user `1001`

Reply:
Updated using long options.

Reply:
@dtrifiro sry, I miss understand your question, I thought as you were adding the user to the root group already you knew it was necessary as @kpouget mentioned.